### PR TITLE
[dynamo] re-add option to log all guard check fails

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113585

Followup to https://github.com/pytorch/pytorch/pull/110325 - re-add the `report_all_guard_failures config` as a logging artifact `recompiles_verbose` with the following changes:
- evaluating the check must be wrapped with exception handling because subsequent code parts following the first failure may result in errors if evaluated (e.g. if a guard checks first for size, then tries to index - a guard failure due to insufficient size would result in an index error for the latter check).
- Adding a test for this case

Sample:
```python
import torch

def fn(x):
    return torch.rand(x[-1], len(x))

opt_fn = torch.compile(fn)
opt_fn([4, 5, 6])
opt_fn([7, 8])
opt_fn([9])
```

Output (with `TORCH_LOGS="recompiles_verbose"`):
```bash
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
[2023-11-15 16:13:26,741] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][1] == 5
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] Recompiling function fn in /data/users/williamwen/pytorch/playground5.py:15
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     triggered by the following guard failure(s):
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 0 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 2
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG] 
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     guard 1 failures:
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - len(L['x']) == 3
[2023-11-15 16:13:26,970] torch._dynamo.guards.__recompiles_verbose: [DEBUG]     - L['x'][0] == 4
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/config.py
Comment (jon-chuang) [inline torch/_dynamo/config.py:271]: maybe we can call it `report_all_guard_failure_checks`

comment: `For each guard failure, whether to print all the failed checks of the guard, or only the first one`

**Review by jon-chuang — APPROVED** [review]
LGTM in principle, as long as tests pass. Thanks!

**Comment by jon-chuang** [comment]
@williamwen42 need approval from core maintainer.

cc @ezyang as reviewer of previous related.

**Comment by williamwen42** [comment]
@jon-chuang do you think it would be better if I included an exception for guard checks that error out? e.g. in the example above, did the guard check `L['x'][2] == 6` fail because of an index error or because equality was not satisfied?

**Comment by jon-chuang** [comment]
> do you think it would be better if I included an exception for guard checks that error out

I think it's better to print the exception as it's a verbose log? Just not the stacktrace.

**Comment by ezyang** [comment]
TBH, I wouldn't print the guards that fail with exceptions at all. They are guaranteed to have been adequately covered by earlier guards and are not contributing much.

File: torch/_dynamo/config.py
Comment (ezyang) [inline torch/_dynamo/config.py:273]: I think this should not be a config. Instead, it should be a logging artifact, so it can be operated remotely via TORCH_LOGS

Comment (jon-chuang) [reply torch/_dynamo/config.py:273]: hmm, I guess instead of `report_all_guard_failure_checks` one could check `all_guard_failures_log.isEnabledFor(logging.DEBUG)`?

**Comment by williamwen42** [comment]
@jon-chuang do you have an example where debugging would be easier if one knows which guard checks will fail after the first failure?

File: torch/_dynamo/guards.py
Comment (ezyang) [inline torch/_dynamo/guards.py:1261]: see my comment

**Review by ezyang — APPROVED** [review]
approving in the spirit of not getting bogged down

**Comment by jon-chuang** [comment]
I don't have one off the top of my head.

But sometimes the order of the checks matter (e.g. a len check might fail, but there is some related other check about object identity that is more important to why the guard actually failed.

**Comment by ezyang** [comment]
Sorry, to be clear, I'm OK with reporting all boolean expressions which return False.

But if I have `len(xs) == 2 and foo(xs[1]) and bar(xs[1])`, if xs is actually length 1, xs[1] is going to raise an exception, and I DON'T care about these guard failures, because they like don't matter, the whole point is that xs[1] doesn't exist, don't give me excruciating detail about all the guards on xs[1] when the element doesn't exist in the first place.

**Comment by williamwen42** [comment]
Changes:
- use logging artifact `recompiles_verbose` instead of config
- do not output guard checks that error out

File: torch/_dynamo/guards.py
Comment (jon-chuang) [inline torch/_dynamo/guards.py:1241]: You can just write `recompiles_verbose_log.isEnabledFor(logging.DEBUG)`

Comment (jon-chuang) [reply torch/_dynamo/guards.py:1241]: The log is only enabled if the artifact logger is enabled I believe?

File: torch/_dynamo/guards.py
Comment (ezyang) [inline torch/_dynamo/guards.py:1344]: This is kind of weird, why are you using an if statement on isEnabledFor on recompiles_log here?

Comment (jon-chuang) [reply torch/_dynamo/guards.py:1344]: I think it should all go to verbose if verbose is enabled, and go to recompile if it is enabled. I.e. verbose will override the recompiles.

Comment (jon-chuang) [reply torch/_dynamo/guards.py:1344]: So verbose should be checked first, no?

Comment (williamwen42) [reply torch/_dynamo/guards.py:1344]: I think it would be more complicated to log partially to regular `recompiles` and partially to `recompiles_verbose` - so I'm just picking one of them based on what's enabled.

**Comment by williamwen42** [comment]
- Only check if artifact is enabled, no need to check for the level of the logger object (enabled artifact assumes logger's level is proper)
- Prioritize logging to `recompiles_verbose` over `recompiles` if both are active.
- Fix bug in torch/_dynamo/utils.py: `logging.get_loggers() -> torch._logging.get_loggers()`

File: torch/_dynamo/guards.py
Comment (ezyang) [reply torch/_dynamo/guards.py]: yes

**Pull request closed without merging**
